SPARC Manual Focus

lazem opened the pull request.

Kleijwegt commented on src/odemis/acq/align/autofocus.py:
```diff
@@ -756,6 +756,131 @@ def Sparc2AutoFocus(align_mode, opm, streams=None, start_autofocus=True):
+    """
+    The actual implementation of the manual focus procedure, run asynchronously
+    """
+    try:
```
You forgot to list the input parameters

Kleijwegt commented on src/odemis/acq/align/autofocus.py:
```diff
@@ -756,6 +756,131 @@ def Sparc2AutoFocus(align_mode, opm, streams=None, start_autofocus=True):
+def _DoSparc2ManualFocus(future, turn_on, opm, bl, align_mode, streams):
+    """
+    The actual implementation of the manual focus procedure, run asynchronously
+    """
```
You forgot to list the input parameters

lazem replied:
Thanks, done as requested.

Kleijwegt commented on src/odemis/gui/cont/tabs.py:
```diff
@@ -4025,6 +4180,7 @@ def _onAutofocus(self, active):
                 return
             btn.SetLabel("Auto focus")
 
+
```
unnecessary new line

pieleric commented on src/odemis/acq/align/autofocus.py:
```diff
@@ -756,6 +756,138 @@ def Sparc2AutoFocus(align_mode, opm, streams=None, start_autofocus=True):
+       align_mode (str): OPM mode, spec-focus or spec-fiber-focus, streak-focus
+       opm: OpticalPathManager object
+       bl: bright light object
+       streams: list of streams
```
Drop this `streams` argument, as noted below...

lazem commented on src/odemis/gui/cont/tabs.py:
```diff
@@ -3918,78 +3932,219 @@ def _onClickFocus(self, evt):
+        # Ignore streak align mode..
+        align_mode = self.tab_data_model.align_mode.value
+        if align_mode == "streak-align":
+            return
```
It's taken care of by `self.panel.pnl_focus.Enable(False)` but this check is still needed since both lens and streak focus go through the same manual focus function. 
Moved it to the caller.

lazem commented on src/odemis/gui/main.xrc:
```diff
@@ -2194,43 +2219,62 @@
+					
+					
+					<object class="sizeritem">
+                        <object class="wxStaticText" name = "cmb_focus_detectors_label">
```
Done. Left the gratings label as it's needed in the code.

lazem commented on src/odemis/gui/cont/tabs.py:
```diff
@@ -3918,78 +3932,219 @@ def _onClickFocus(self, evt):
+        # Pick the stream to be shown:
+        #  1. Pick the stream which is playing (should_update=True)
+        #  2. Pick the focus stream which is already shown (in v.getStreams())
+        #  3. Pick the first focus stream
```
Listening now to the focus stream should_update directly without closure.

pieleric commented on src/odemis/acq/align/autofocus.py:
```diff
@@ -756,6 +756,135 @@ def Sparc2AutoFocus(align_mode, opm, streams=None, start_autofocus=True):
+                logging.info("Manual focus procedure cancelled before the light is on")
+                raise CancelledError()
+            bl.power.value = bl.power.range[1]
+            bl.emissions.value = [1] * len(bl.emissions.value)
```
Actually, you don't need to wait for the path to be ready before turning on the light. You can do it at the same time. So in-between `opm.setPath(align_mode)` and `.result()`. In this case, you only have one step (of 10s).

lazem replied:
Done with the suggested refactoring. Just kept the canceller so at least the progress bar appears stopped (even if the opm future didn't get cancelled).

pieleric commented on src/odemis/acq/align/autofocus.py:
```diff
@@ -756,6 +756,135 @@ def Sparc2AutoFocus(align_mode, opm, streams=None, start_autofocus=True):
+
+    return True
+
+def Sparc2AManualFocus(turn_on, align_mode, opm, bl):
```
Let's drop the `A`, -> `Sparc2ManualFocus()`

pieleric commented:
> I have somewhat the feeling that some corner cases might not be properly handled (eg, cancel manual focus on streak cam). Admittedly that's quite a bit of a pain to test everything (manually). Make sure to be extra careful.

Sorry, I realise that actually when switching to manual focus, it's not really possible to cancel, unless changing mode or tab. So it might be a little bit easier to handle.

Kleijwegt commented on src/odemis/gui/cont/tabs.py:
```diff
@@ -3202,6 +3204,7 @@ def __init__(self, name, button, panel, main_frame, main_data):
         self._spot_stream = spot_stream
 
         self._ccd_stream = None
+        self._ccd_spe = None
```
What does SPE stand for?

lazem replied:
Good question! This is a stream controller for the ccd stream, it's initialized in here 

```
self._ccd_spe = self._stream_controller.addStream(ccd_stream,
                                add_to_view=self.panel.vp_align_lens.view)
```

Maybe @pieleric can tell us, I used it as it's defined.

Kleijwegt replied:
Maybe it is nice to add a comment with the explanation behind the initialization of this variable?

pieleric replied:
Stream Panel Entry.

lazem replied:
Added a comment for the reason of keeping it as attribute

Kleijwegt commented on src/odemis/gui/cont/tabs.py:
```diff
@@ -3313,6 +3286,36 @@ def __init__(self, name, button, panel, main_frame, main_data):
             self.panel.pnl_fib_focus.Show(False)
 
+        # Manual focus mode initialization
+        # All the `blue` streams one for each detector to adjust for focus
```
Seems like in this sentence a typo slipped in? Maybe a comma need to be add? Maybe the word All needs to be Add? What do you mean exactly?

Kleijwegt commented on src/odemis/gui/cont/tabs.py:
```diff
@@ -3918,78 +3922,188 @@ def _onClickFocus(self, evt):
+                         model.MD_ROTATION: 0},
+                acq_type=model.MD_AT_SLIT,
+            )
+            speclines.tint.value = (0, 64, 255)  # colour it blue
```
Maybe the imported blue should be a color defined in 

> src/odemis/gui/__init__.py

So maybe import from this place or reference to it in a comment

lazem replied:
Done as requested.

Kleijwegt commented on src/odemis/gui/cont/tabs.py:
```diff
@@ -3918,78 +3922,188 @@ def _onClickFocus(self, evt):
+        stream.should_update.value = True
+        stream.is_active.value = True
+
+    def add_combobox_control(self, label_text, value=None, conf=None):
```
Maybe this method should be private?

lazem replied:
It needs to be the same signature to be used correctly by `create_axis_entry`.

lazem commented on src/odemis/acq/align/autofocus.py:
```diff
@@ -757,6 +757,82 @@ def Sparc2AutoFocus(align_mode, opm, streams=None, start_autofocus=True):
     return f
 
 
+def _cancelSparc2ManualFocus(future):
```
All methods for focus are inside autofocus.py, the public ones are `Sparc2AutoFocus` and `Sparc2ManualFocusused`, used by `Sparc2AlignTab` object. The rest is private.

pieleric commented on src/odemis/gui/cont/tabs.py:
```diff
@@ -3918,78 +3922,188 @@ def _onClickFocus(self, evt):
+            # Update GUI
+            self._pfc_manual_focus = ProgressiveFutureConnector(self._mf_future, gauge)
+
+    def _onManualFocusReady(self, future):
```
You should protect this function with a `@call_in_wx_main`, as it will be run in a separate thread and it accesses the GUI.

pieleric commented on src/odemis/gui/cont/tabs.py:
```diff
@@ -4003,6 +4117,8 @@ def _onAutofocus(self, active):
             # GUI stream bar controller pauses the stream
             btn.SetLabel("Cancel")
+            self.panel.btn_manual_focus.Enable(False)
+            self._enableFocusComponents(ccd_stream=False)
```
Please explicitly write `manual=False`

pieleric commented on src/odemis/gui/cont/tabs.py:
```diff
@@ -3918,78 +3922,188 @@ def _onClickFocus(self, evt):
+        #  3. Pick the first focus stream
+        """
+        focusedview = self.tab_data_model.focussedView.value
+        should_update_stream = next((s for s in self._focus_streams if s.should_update.value == True), None)
```
Drop the `== True`. Booleans can be directly used as-is.

pieleric commented on src/odemis/gui/cont/tabs.py:
```diff
@@ -3918,78 +3922,188 @@ def _onClickFocus(self, evt):
+            # Set the focus detectors combobox selection
+            try:
+                istream = self._focus_streams.index(should_update_stream)
+                if istream != -1:
```
Drop this `if`. `index()` always returns >= 0.

lazem commented on src/odemis/acq/align/autofocus.py:
```diff
@@ -757,6 +757,82 @@ def Sparc2AutoFocus(align_mode, opm, streams=None, start_autofocus=True):
+        future._running_subf.cancel()
+        logging.debug("Sparc2AManualFocus cancellation requested.")
+
+    return True
```
That's right, it keeps going when the return is not True.

lazem commented on src/odemis/gui/cont/tabs.py:
```diff
@@ -3918,78 +3922,188 @@ def _onClickFocus(self, evt):
+        # No need to create components, defined already on the xrc file
+        return self.panel.cmb_focus_gratings_label, self.panel.cmb_focus_gratings
+
+    def _enableFocusComponents(self, manual=False, ccd_stream=False):
```
Dropped default value of `manual`.

pieleric commented on src/odemis/gui/cont/tabs.py:
```diff
@@ -3918,78 +3924,188 @@ def _onClickFocus(self, evt):
+
+        return focus_streams
+
+    def _ensureOneFocusStream(self, _):
```
Should be also protected by `@call_in_wx_main`.

lazem commented:
> Looks fine. There is just one extra change that I think is needed.
> Can you add the suggested line, and squash the commits?

Added the suggested line and squashed the commits. 
@pieleric
